### PR TITLE
Feature/44 make flash attention configurable

### LINKED ISSUE #36: Make Flash attention configurable
### Is your feature request related to a problem? Please describe.

The current implementation causes issues when loading old model checkpoints during inference as it is not clear whether flash attention was used or not.

### Describe the solution you'd like

A class MultiHeadSelfAttention and FlashMultiHeadSelfAttention which inherits from the former, but makes use of flash attention. This should be set by the user in the config. 

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### Organisation

ECMWF

### PR DESCRIPTION
**Pull request by theissenhelen**
Current setup:
- If flash-attn is available in the environment, the MultiHeadSelfAttention module automatically imports the corresponding attention function. In inference however we do not have that information.

Now:
- flex attention available
- user specifies whether flash-attn, flex attention or scaled dot product attention should be used in the model config.
- adds configurable parameters (soft cap, aLiBi) for flash attention
- for aLiB:i adds a function to compute the slopes according to the number of attention heads
- scaled dot product attention now supports sliding window (making it numerically equivalent to flash/flex)

This PR will be accompanied by changes to the config in Anemoi-training ([PR](https://github.com/ecmwf/anemoi-training/pull/73))

Todo:
- [x] test various attention options

----
📚 Documentation preview 📚: https://anemoi-models--47.org.readthedocs.build/en/47/

### CODE REVIEW

File: src/anemoi/models/layers/attention.py [hunk 1]
```diff
@@ -11,19 +11,11 @@
 from typing import Optional
 
 import einops
+from flash_attn import flash_attn_func
```
Comment (mchantry) [inline src/anemoi/models/layers/attention.py:14]: When is the import triggered? Only when building models with attention?

Comment (theissenhelen) [reply src/anemoi/models/layers/attention.py:14]: good point, I will put into the init instead

File: CHANGELOG.md [hunk 2]
```diff
@@ -27,15 +27,23 @@ Keep it human-readable, your future self will thank you!
 
 ### Added
 
+- CI workflow to update the changelog on release
```
Comment (mchantry) [inline CHANGELOG.md:30]: Tag #47 please on these lines

File: src/anemoi/models/layers/attention.py [hunk 3]
```diff
@@ -41,31 +36,74 @@ def __init__(
...
-        if not _FLASH_ATTENTION_AVAILABLE:
-            LOGGER.warning("Flash attention not available, falling back to pytorch scaled_dot_product_attention")
+        self.lin_qkv = nn.Linear(embed_dim, 3 * embed_dim, bias=bias)
 
         self.projection = nn.Linear(embed_dim, embed_dim, bias=True)
 
+    def set_attention_function(self):
+
+        if self.use_flash_attention:
+            from flash_attn import flash_attn_func
```
Comment (mpvginde) [inline src/anemoi/models/layers/attention.py:95]: Not sure if really necessary, but might be useful to introduce a version check.
I was using an older version of flash_attn and there the `flash_attn_func` had no `window_size` keyword, resulting in crash. 
Or you could enforce the version via de optional dependencies.

Comment (theissenhelen) [reply src/anemoi/models/layers/attention.py:95]: Flash-attn and pyproject.toml does not work well together in the current setup, so I think a check would be the better option.

**Review by mpvginde — DISMISSED** [review]

File: src/anemoi/models/layers/attention.py [hunk 4]
```diff
@@ -88,11 +130,23 @@ def forward(
         value = shard_heads(value, shapes=shapes, mgroup=model_comm_group)
         dropout_p = self.dropout_p if self.training else 0.0
 
-        if _FLASH_ATTENTION_AVAILABLE:
+        if self.use_flash_attention:
```
Comment (b8raoult) [inline src/anemoi/models/layers/attention.py:133]: Is this boolean persistant in the checkpoint?

Comment (b8raoult) [reply src/anemoi/models/layers/attention.py:133]: And also, why do we test it in `forward` when we used it in `__init__`?

Comment (theissenhelen) [reply src/anemoi/models/layers/attention.py:133]: It should be persistent and it is also part of the model metadata.

We need it in the init to set the attention function and we need it in the forward because we perform slightly different operations for the two attention implementations.

File: CHANGELOG.md [hunk 5]
```diff
@@ -27,15 +27,24 @@ Keep it human-readable, your future self will thank you!
...
+- add configurability of flash attention (#47)
+- configurabilty of the dropout probability in the the MultiHeadSelfAttention module
 - CI workflow to update the changelog on release
 - Remapper: Preprocessor for remapping one variable to multiple ones. Includes changes to the data indices since the remapper changes the number of variables. With optional config keywords.
+- Codeowners file
+- Pygrep precommit hooks
+- Docsig precommit hooks
+- Changelog merge strategy
+
```
Comment (HCookie) [inline CHANGELOG.md:39]: These log entries probably shouldn't be a part of this PR.

Comment (theissenhelen) [reply CHANGELOG.md:51]: If I remove them, the Changelog will be incorrect.

**Review by b8raoult — DISMISSED** [review]

**Review by b8raoult — DISMISSED** [review]

**Comment by FussyDuck** [comment]
[CLA assistant check](https://bol-claassistant.ecmwf.int:443/ecmwf/anemoi-models?pullRequest=47) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://bol-claassistant.ecmwf.int:443/ecmwf/anemoi-models?pullRequest=47) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: theissenhelen<br/>:x: cathalobrien<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://bol-claassistant.ecmwf.int:443/check/ecmwf/anemoi-models?pullRequest=47) it.</sub>

File: src/anemoi/models/layers/attention.py [hunk 6]
```diff
@@ -42,31 +38,85 @@ def __init__(
...
         self.projection = nn.Linear(embed_dim, embed_dim, bias=True)
 
+    def set_attention_function(self):
+        attn_funcs = {
+            "flash_attention": FlashAttentionWrapper,
+            "flex_attention": FlexAttentionWrapper,
+            "scaled_dot_product_attention": TorchAttentionWrapper,
+        }
+        if self.attention_implementation in attn_funcs:
+            LOGGER.info(f"attention.py: using {self.attention_implementation}")
```
Comment (JesperDramsch) [inline src/anemoi/models/layers/attention.py:106]: Should this actually name the file?
I think this is the job of the logger and increases clutter. 

I would change this throughout the file, actually.

File: src/anemoi/models/layers/attention.py [hunk 7]
```diff
@@ -42,31 +38,85 @@ def __init__(
...
+        window_size : Optional[int], optional
+            window_size, by default None
+        dropout_p : float, optional
+            dropout probability, by default 0.0
+        softcap : float, optional
+            Anything > 0 activates softcapping attention, by default None
+        attention_implementation: str, optional
+            A predefined string which selects which underlying attention
+            implementation, by default "flash_attention"
+        use_alibi_slopes : bool, optional
```
Comment (JesperDramsch) [inline src/anemoi/models/layers/attention.py:66]: nit: I wonder if the explanation of the alibit slope should rather be in the main part of the docstring to keep this a bit leaner.

File: src/anemoi/models/layers/attention.py [hunk 8]
```diff
@@ -42,31 +38,85 @@ def __init__(
...
+        self.window_size = window_size
         self.dropout_p = dropout_p
         self.is_causal = is_causal
+        self.softcap = softcap
 
-        self.lin_qkv = nn.Linear(embed_dim, 3 * embed_dim, bias=bias)
-        self.attention = attn_func
+        if self.use_alibi_slopes:
+            self.alibi_slopes = get_alibi_slopes(num_heads)
+            assert self.alibi_slopes.shape[0] == num_heads
```
Comment (JesperDramsch) [inline src/anemoi/models/layers/attention.py:91]: This should have a message. I don't know what it's testing and would have no idea how to debug.

File: src/anemoi/models/layers/attention.py [hunk 9]
```diff
@@ -89,24 +139,220 @@ def forward(
...
+        super().__init__()
+
+        from torch.nn.functional import scaled_dot_product_attention
+
+        self.attention = scaled_dot_product_attention
+        self.mask = None
+        self.window_size = None
+
+    def update_mask(self, seq_len, window_size: int, device: str):
+        update_mask = (
```
Comment (JesperDramsch) [inline src/anemoi/models/layers/attention.py:175]: nit: why is this saved as variable?

File: src/anemoi/models/layers/attention.py [hunk 10]
```diff
@@ -89,24 +139,220 @@ def forward(
...
+            self.mask is None or self.window_size != window_size or tuple(self.mask.shape) != (seq_len, seq_len)
```
Comment (JesperDramsch) [inline src/anemoi/models/layers/attention.py:176]: This could use a comment to explain the logic.

I find it a bit weird that we call `update_mask` and sometimes it doesn't update the mask without informing the user in any way.

File: src/anemoi/models/layers/attention.py [hunk 11]
```diff
@@ -89,24 +139,220 @@ def forward(
...
+        value,
+        batch_size: int,
+        causal=False,
+        window_size=None,
+        dropout_p=0.0,
+        softcap=None,
+        alibi_slopes=None,
+    ):
+        if softcap is not None:
+            SystemError(
```
Comment (JesperDramsch) [inline src/anemoi/models/layers/attention.py:201]: Shouldn't this be a `NotImplementedError`?

File: src/anemoi/models/layers/attention.py [hunk 12]
```diff
@@ -89,24 +139,220 @@ def forward(
...
+                "Error. Softcap not supported by Pytorchs SDPA. please switch to flash attention or disable softcap."
```
Comment (JesperDramsch) [inline src/anemoi/models/layers/attention.py:202]: The error should not say it's an error. This increases clutter.

File: src/anemoi/models/layers/attention.py [hunk 13]
```diff
@@ -89,24 +139,220 @@ def forward(
...
+            )
+        if alibi_slopes is not None:
+            SystemError(
+                "Error. Alibi slopes not supported by Pytorchs SDPA. please switch to flash attention or disable alibi slopes."
+            )
+        if window_size is not None:
```
Comment (JesperDramsch) [inline src/anemoi/models/layers/attention.py:208]: Ok, so we're already checking if we should update the mask and then in the `update_mask` we do a bunch of other checks. I think this needs a small refactor to be more tidy.

File: src/anemoi/models/layers/attention.py [hunk 14]
```diff
@@ -89,24 +139,220 @@ def forward(
...
+            self.update_mask(query.shape[-2], window_size=window_size, device=query.device)
         else:
+            self.mask = None
+
+        with torch.nn.attention.sdpa_kernel(backends=[torch.nn.attention.SDPBackend.MATH]):
```
Comment (JesperDramsch) [inline src/anemoi/models/layers/attention.py:213]: What is the reasoning to limit the backend?

File: src/anemoi/models/layers/attention.py [hunk 15]
```diff
@@ -89,24 +139,220 @@ def forward(
...
-        out = self.projection(out)
+
+class FlexAttentionWrapper(nn.Module):
+    """Wrapper for Pytorch Flex attention."""
+
+    def __init__(self):
+        super().__init__()
+
+        if version.parse(torch.__version__) < version.parse("2.5.0"):
+            raise SystemExit("Error: torch version is too low. Update to 2.5.0 or higher to use Flex Attention.")
```
Comment (JesperDramsch) [inline src/anemoi/models/layers/attention.py:233]: Wrong error

File: src/anemoi/models/layers/attention.py [hunk 16]
```diff
@@ -89,24 +139,220 @@ def forward(
...
+            self.block_mask = create_block_mask(
+                sliding_window_mask, B=None, H=None, Q_LEN=seq_len, KV_LEN=seq_len, _compile=True
+            )
+            self.attention = functools.partial(
+                flex_attention, block_mask=self.block_mask
+            )  # Cache the block mask (recomended in attn blog post)
+            self.attention = torch.compile(self.attention)
+            self.is_attn_compiled = True
+
+        # TODO test how this impacts scaling at large model counts
```
Comment (JesperDramsch) [inline src/anemoi/models/layers/attention.py:286]: Who is this a TODO for?

File: src/anemoi/models/layers/attention.py [hunk 17]
```diff
@@ -89,24 +139,220 @@ def forward(
...
+    ----------
+    num_heads : int
+        number of attention heads
+
+    Returns
+    -------
+    Tensor
+        aLiBi slopes
+    """
+    n = 2 ** math.floor(math.log2(num_heads))
```
Comment (JesperDramsch) [inline src/anemoi/models/layers/attention.py:351]: Since `num_heads` is an integer, we could be using bit-shifting here:
`n = 1 << (num_heads.bit_length() - 1)`

Not sure how necessary speed is here though, as a trade-off against readability. It would definitely need a comment.

File: src/anemoi/models/layers/attention.py [hunk 18]
```diff
@@ -89,24 +139,220 @@ def forward(
...
+    slope_0 = 2.0 ** (-8.0 / n)
```
Comment (JesperDramsch) [inline src/anemoi/models/layers/attention.py:352]: Could be simplified as Python doesn't need explicit float typing to: `2 ** (- 8 / n)`

File: src/anemoi/models/layers/attention.py [hunk 19]
```diff
@@ -89,24 +139,220 @@ def forward(
...
+    alibi_slopes = torch.pow(slope_0, torch.arange(1, 1 + n))
+    if n < num_heads:
+        slope_hat_0 = 2.0 ** (-4.0 / n)
```
Comment (JesperDramsch) [inline src/anemoi/models/layers/attention.py:355]: Could be simplified as Python doesn't need explicit float typing to: `2 ** (-4/n)`.

But realistically, `slope_hat_0` is just `slope_0 ** .5`.

File: src/anemoi/models/layers/chunk.py [hunk 20]
```diff
@@ -92,6 +95,13 @@ def __init__(
             Activation function, by default "GELU"
         dropout_p: float
             Dropout probability used for multi-head self attention, default 0.0
+        attention_implementation: str, optional
+            A predefined string which selects which underlying attention
+            implementation, by default "flash_attention"
+        softcap : float, optional
+            Anything > 0 activates softcapping flash attention, by default None
```
Comment (JesperDramsch) [inline src/anemoi/models/layers/chunk.py:102]: What does "Anything > 0" mean here? Please adjust this explanation across docstrings to be more informative to someone that hasn't worked with the attention implementation yet.

**Review by JesperDramsch — CHANGES_REQUESTED** [review]
Thanks for this contribution. This should make the code much more stable in inference.

I left a few comments about implementation details that caught my eye and the documentation needs to be updated.

Additionally, now that it's configurable, is there a way to also change it in the config?
I assume it's through the instantiation. Should we add it to the config then to make it explicitly available?

When the errors are implemented correctly, please also add tests that make sure the errors are triggered correctly, so we can catch edge-cases.

File: src/anemoi/models/layers/attention.py (hunk 14 above)
Comment (japols) [reply src/anemoi/models/layers/attention.py:218]: @JesperDramsch the idea is to use the `MATH` backend because it comes with pytorch and doesn't rely on external libraries like the other two options:

* `FLASH_ATTENTION` which we allready have as an explicit option, or 
* `EFFICIENT_ATTENTION` [memory-efficient-attention](https://github.com/facebookresearch/xformers)

I would rather support the last one explicitly than switching via SDPA backend

File: src/anemoi/models/layers/attention.py (hunk 17 above)
Comment (theissenhelen) [reply src/anemoi/models/layers/attention.py:356]: Speed is not an issue as it is only calculated once.
